feat(go): introduce standalone sample

stefanoamorelli opened the pull request:
> [!TIP]
> Better reviewed commit-by-commit.

# Description

Implements a standalone `go` sample.

- [x] Follow the [`CONTRIBUTING` Guide](https://github.com/google-agentic-commerce/AP2?tab=contributing-ov-file#how-to-contribute).
- [x] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [x] Ensure the tests and linter pass
- [x] Appropriate docs were updated (if necessary)

Closes #71  🦕

stefanoamorelli commented on samples/go/pkg/roles/merchant_agent/executor.go:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	agentIDStr := fmt.Sprintf("%v", agentID)
+	log.Printf("Received request from shopping_agent_id: %s", agentIDStr)
```
Done - using type assertion with error handling

stefanoamorelli commented on samples/go/README.md:
```diff
@@ -0,0 +1,237 @@
+│       ├── credentials_provider_agent/
+│       └── merchant_payment_processor_agent/
+├── go.mod
+└── README.md
```
Done - cleaned up diagram with separate explanation

stefanoamorelli commented on samples/go/pkg/common/function_resolver.go:
```diff
@@ -0,0 +1,177 @@
+		}
+	}
+	return false
+}
```
Done - using strings.Contains()

stefanoamorelli commented on samples/go/pkg/roles/merchant_agent/storage.go:
```diff
@@ -0,0 +1,150 @@
+	defer s.mutex.RUnlock()
+
+	return s.products
+}
```
Done - added TODO comment

joshlund-goog commented on samples/go/go.sum:
```diff
@@ -0,0 +1,207 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
/usr/local/google/home/joshlund/go/pkg/mod/go.opentelemetry.io/otel@v1.37.0/internal/global/trace.go:28:2: missing go.sum entry for module providing package go.opentelemetry.io/auto/sdk (imported by go.opentelemetry.io/otel/internal/global); to add:
	go get go.opentelemetry.io/otel/internal/global@v1.37.0
make: *** [Makefile:22: build] Error 1

I verified `go get go.opentelemetry.io/otel/internal/global@v1.37.0` fixes

joshlund-goog commented on samples/go/README.md:
```diff
@@ -0,0 +1,319 @@
+
+### Cross-Language Integration (Recommended)
+
+**1. Start Go backend agents** (as shown above)
```
Please add a note that the following steps will be done in a separate terminal.

joshlund-goog commented on samples/go/pkg/common/server.go:
```diff
@@ -0,0 +1,129 @@
+
+func (s *AgentServer) setupRoutes() {
+	s.router.HandleFunc(s.RPCURL, s.handleA2ARequest).Methods("POST")
+	s.router.HandleFunc(s.RPCURL+"/card", s.handleGetCard).Methods("GET")
```
Agent cards should be hosted at `https://{agent-server-domain}/.well-known/agent-card.json` 
source: https://a2a-protocol.org/latest/topics/agent-discovery/#1-well-known-uri

This will fix issues integrating with the adk-python Shopping Agent:
HTTP Error 404: Failed to fetch agent card from http://localhost:8001/a2a/merchant_agent/.well-known/agent-card.json

Please update your server logging and http_client.go to match

joshlund-goog commented:
> > Could you add a README that instructs the user:
> > ```
> > * how to run this sample
> > 
> > * how to interact with this sample (I notice there is no shopping_agent implemented in go)
> > 
> > * what the sample is demonstrating
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > An example may be found at samples/python/scenarios/a2a/human-present/cards/README.md.
> 
> Thanks for the feedback @joshlund-goog! I've updated the READMEs to address your points, but I also wanted to ask your opinion on something before finalizing this. Right now the `golang` sample focuses on the three backend agents (merchant, credentials provider, payment processor) without a `Shopping Agent` because the Python sample uses `adk-python` which handles all the agent orchestration, state management, and the web UI. We don't have an equivalent framework in Go, so I'd need to build a simplified version from scratch. It would work, but it wouldn't have the same polish as the Python version with ADK's dev-ui.
> 
> Alternatively, we can demonstrate the `golang` agents orchestrated using the Python `Shopping Agent`, implemented in `adk-python`. This would show interoperability but it's not a 100% `golang` sample.
> 
> What do you think makes more sense, or do you have other ideas? Please let me know if I'm missing out something.

I like your solution of keeping the adk-python Shopping Agent as the entry point. I sent a review of the snags I ran into testing it out. Please fix and I'll continue testing/reviewing the go sample.

stefanoamorelli commented on samples/go/README.md:
```diff
@@ -0,0 +1,319 @@
+
+## Overview
+
+This directory contains Go implementations of backend agents for the Agent Payments Protocol, demonstrating how to build production-quality AP2 services using Go. These agents implement the merchant-side, payment processing, and credentials provider roles in an AP2 transaction flow.
```
Done :heavy_check_mark:

stefanoamorelli commented on samples/go/README.md:
````diff
@@ -0,0 +1,319 @@
+**3. Start Python Shopping Agent:**
+
+```sh
+cd samples/python
````
:heavy_check_mark:

stefanoamorelli commented:
Thanks @joshlund-goog, I've addressed your comments and added a few improvements.

Here's also an E2E walkthrough of the process: 

https://github.com/user-attachments/assets/ade0c22f-6a47-400a-8436-7d60c4ba6f2b

joshlund-goog commented on samples/python/src/common/payment_remote_a2a_client.py:
```diff
@@ -14,107 +14,113 @@
+    async def get_agent_card(self) -> a2a_types.AgentCard:
+        """Get agent card."""
+        if self._agent_card is None:
+            parsed_url = urlparse(self._base_url)
```
Acknowledging that base_url was not the best name here, but this will likely break the python samples. Their agent card is hosted at the url containing the paths.

Could you update the go sample the host the agent_cards at the same location as the python sample. For example, the well known address at https://github.com/google-agentic-commerce/AP2/blob/f404a8ddf2e2a5ddf76a6f3fd990b4cb78a71200/samples/python/src/roles/merchant_agent/agent.json#L4 

Then revert changes to this file.

stefanoamorelli replied:
@joshlund-goog indeed :pray:  I've dropped the changes on `payment_remote_a2a_client.py` and adjusted the `go` sample accordingly: [`669f2a6` (#101)](https://github.com/google-agentic-commerce/AP2/pull/101/commits/669f2a68e327c1b717cf6ddcac63ed0fd59c062d) in a separate commit for easier review (lmk if I should interactive rebase -> amend, or we can squash on merge).

Also ran another E2E test, successful :heavy_check_mark:

joshlund-goog reviewed (approved):
Verified the E2E test works. Nice work. Couple of README updates and then we can squash on merge.

joshlund-goog commented on .github/workflows/linter.yaml:
```diff
@@ -30,6 +30,8 @@ jobs:
           VALIDATE_PYTHON_ISORT: false
           VALIDATE_PYTHON_MYPY: false
           VALIDATE_PYTHON_PYLINT: false
+          VALIDATE_PYTHON_RUFF: false
```
I believe you can leave the python linters on now.

joshlund-goog commented on samples/go/scenarios/a2a/human-present/cards/README.md:
```diff
@@ -0,0 +1,282 @@
+
+1. **Start the Go backend agents** (see [Quick Start](#quick-start))
+
+2. **Start the Python Shopping Agent:**
```
add small note that this is in a separate terminal

stefanoamorelli commented:
Adjusted as suggested :heavy_check_mark: @joshlund-goog
